File: src/Shell/HelpFormatter.cpp
/*
** EPITECH PROJECT, 2026
** Shell
** File description:
** HelpFormatter
*/

#include "HelpFormatter.hpp"

#include <algorithm>
#include <iomanip>
#include <ranges>
#include <sstream>

namespace shell {

static constexpr std::size_t MAX_WIDTH = 80;
static constexpr std::size_t INDENT    = 4;
static constexpr std::size_t COL_GAP   = 4;

static std::string wrapText(const std::string &text, std::size_t startCol)
{
    if (text.empty())
        return "\n";

    const std::string cont(startCol, ' ');
    std::istringstream words(text);
    std::string result;
    std::string word;
    std::size_t col = startCol;
    bool first = true;

    while (words >> word) {
        if (first) {
            result += word;
            col += word.size();
            first = false;
        } else if (col + 1 + word.size() > MAX_WIDTH) {
            result += "\n" + cont + word;
            col = startCol + word.size();
        } else {
            result += " " + word;
            col += 1 + word.size();
        }
    }
    return result + "\n";
}

static std::string formatSection(
    const std::string &title,
    const std::vector<std::pair<std::string, std::string>> &entries)
{
    if (entries.empty())
        return "";

    std::size_t maxLabelLen = 0;
    for (const auto &label: entries | std::views::keys)
        maxLabelLen = std::max(maxLabelLen, label.size());

    const std::size_t descStart = INDENT + maxLabelLen + COL_GAP;

    std::string result = "\n" + title + "\n";
    for (const auto &[label, desc] : entries) {
        std::ostringstream oss;
        oss << std::string(INDENT, ' ')
            << std::left << std::setw(static_cast<int>(maxLabelLen + COL_GAP))
            << label;
        result += oss.str() + wrapText(desc, descStart);
    }
    return result;
}

static void appendUsageToken(std::string &line, const std::string &token,
    const std::size_t contIndent)
{
    if (line.size() + 1 + token.size() > MAX_WIDTH)
        line += "\n" + std::string(contIndent, ' ') + token;
    else
        line += " " + token;
}

std::string formatHelp(const command::CommandDefinition &def)
{
    const std::size_t usageCont = INDENT + def.name.size() + 1;
    std::string usageLine = std::string(INDENT, ' ') + def.name;

    for (const auto &option : def.options) {
        std::string token = option.required ? "--" : "[--";
        token += option.name;
        if (!option.alias.empty())
            token += " | -" + option.alias;
        if (!option.required)
            token += "]";
        token += " <" + option.name + ">";
        appendUsageToken(usageLine, token, usageCont);
    }

    for (const auto &flag : def.flags) {
        std::string token = "[--" + flag.name;
        if (!flag.alias.empty())
            token += " | -" + flag.alias;
        token += "]";
        appendUsageToken(usageLine, token, usageCont);
    }

    for (const auto &arg : def.arguments) {
        const std::string token = arg.required
            ? "<" + arg.name + ">"
            : "[" + arg.name + "]";
        appendUsageToken(usageLine, token, usageCont);
    }

    std::string result = "Usage\n" + usageLine + "\n";

    if (!def.description.empty()) {
        result += "\nDescription\n";
        result += std::string(INDENT, ' ') + wrapText(def.description, INDENT);
    }

    std::vector<std::pair<std::string, std::string>> argEntries;
    for (const auto &arg : def.arguments)
        argEntries.emplace_back(arg.name, arg.description);
    result += formatSection("Argument(s)", argEntries);

    std::vector<std::pair<std::string, std::string>> optEntries;
    for (const auto &opt : def.options) {
        std::string label = "--" + opt.name;
        if (!opt.alias.empty())
            label += ", -" + opt.alias;
        optEntries.emplace_back(label, opt.description);
    }
    result += formatSection("Option(s)", optEntries);

    std::vector<std::pair<std::string, std::string>> flagEntries;
    for (const auto &flag : def.flags) {
        std::string label = "--" + flag.name;
        if (!flag.alias.empty())
            label += ", -" + flag.alias;
        flagEntries.emplace_back(label, flag.description);
    }
    result += formatSection("Flag(s)", flagEntries);

    return result;
}

std::string formatCommandList(
    const std::unordered_map<std::string, command::CommandDefinition> &commands)
{
    std::size_t maxLen = 0;
    for (const auto &name : commands | std::views::keys)
        maxLen = std::max(maxLen, name.size());

    const std::size_t descStart = INDENT + maxLen + COL_GAP;

    std::string result = "Available commands:\n";
    for (const auto &[name, def] : commands) {
        std::ostringstream oss;
        oss << std::string(INDENT, ' ')
            << std::left << std::setw(static_cast<int>(maxLen + COL_GAP))
            << name;
        result += oss.str() + wrapText(def.description, descStart);
    }
    return result;
}

} // namespace shell

File: src/AIClient/shell/src/command/help_formatter.rs
use crate::command::CommandDefinition;
use std::cmp::max;
use std::collections::HashMap;

const MAX_WIDTH: usize = 80;
const INDENT: usize = 4;
const COL_GAP: usize = 4;

fn wrap_text(text: &str, start_col: usize) -> String {
    match text.is_empty() {
        true => "\n".to_string(),
        false => {
            let indent = " ".repeat(start_col);
            let words = text.split_whitespace();
            let mut result = String::new();
            let mut len = start_col;
            let mut first = true;

            for word in words {
                if first {
                    result.push_str(&word);
                    len += word.len();
                    first = false;
                } else if len + 1 + word.len() > MAX_WIDTH {
                    result.push_str(&format!("\n{indent}{word}"));
                    len = start_col + word.len();
                } else {
                    result.push_str(&format!(" {word}"));
                    len += 1 + word.len();
                }
            }

            result + "\n"
        }
    }
}

fn format_section(title: &str, entries: &[(String, String)]) -> String {
    match entries.is_empty() {
        true => String::new(),
        false => {
            let mut max_label_length = 0;
            for (label, _) in entries {
                max_label_length = max(max_label_length, label.len());
            }

            let description_start = INDENT + max_label_length + COL_GAP;
            let mut result = format!("\n{title}\n");

            let indent = " ".repeat(INDENT);
            for (label, description) in entries {
                result.push_str(&format!(
                    "{indent}{label:<width$}{}",
                    wrap_text(description, description_start),
                    width = max_label_length + COL_GAP,
                ));
            }

            result
        }
    }
}

fn append_usage(line: &mut String, token: String, indent_size: usize) {
    match line.len() + 1 + token.len() > MAX_WIDTH {
        true => line.push_str(&format!("\n{}{token}", " ".repeat(indent_size))),
        false => line.push_str(&format!(" {token}")),
    }
}

fn build_usage(command: &CommandDefinition) -> String {
    let indent_size = INDENT + command.name.len() + 1;
    let mut usage = format!("{}{}", " ".repeat(INDENT), command.name);

    for arg in &command.arguments {
        append_usage(
            &mut usage,
            if arg.required {
                format!("<{}>", arg.name)
            } else {
                format!("[{}]", arg.name)
            },
            indent_size,
        )
    }

    for option in &command.options {
        let mut token = if option.required {
            "--".to_string()
        } else {
            "[--".to_string()
        };
        token.push_str(&option.name);
        if let Some(alias) = &option.alias {
            token.push_str(&format!(" | -{}", alias))
        }
        if !option.required {
            token.push_str("]");
        }
        token.push_str(&format!(" <{}>", option.name));
        append_usage(&mut usage, token, indent_size);
    }

    for flag in &command.flags {
        append_usage(
            &mut usage,
            format!(
                "[--{}{}]",
                flag.name,
                if let Some(alias) = &flag.alias {
                    format!(" | -{}", alias)
                } else {
                    "".to_string()
                }
            ),
            indent_size,
        )
    }

    usage.push_str(&format!("\n{}{}", " ".repeat(INDENT), command.name));
    append_usage(&mut usage, "--help | -h".to_string(), indent_size);

    usage
}

pub fn format_help(command: &CommandDefinition) -> String {
    let mut result = format!("Usage\n{}\n", build_usage(command));
    let indent = " ".repeat(INDENT);

    if !command.description.is_empty() {
        result.push_str(&format!(
            "\nDescription\n{indent}{}",
            wrap_text(&command.description, INDENT)
        ))
    }

    let arg_entries: Vec<(String, String)> = command
        .arguments
        .iter()
        .map(|arg| (arg.name.clone(), arg.description.clone()))
        .collect();
    result.push_str(&format_section("Argument(s)", &arg_entries));

    let option_entries: Vec<(String, String)> = command
        .options
        .iter()
        .map(|option| {
            (
                format!(
                    "--{}{}",
                    option.name,
                    if let Some(alias) = &option.alias {
                        format!(", -{alias}")
                    } else {
                        String::new()
                    }
                ),
                option.description.clone(),
            )
        })
        .collect();
    result.push_str(&format_section("Option(s)", &option_entries));

    let flag_entries: Vec<(String, String)> = command
        .flags
        .iter()
        .map(|flag| {
            (
                format!(
                    "--{}{}",
                    flag.name,
                    if let Some(alias) = &flag.alias {
                        format!(", -{}", alias)
                    } else {
                        String::new()
                    }
                ),
                flag.description.clone(),
            )
        })
        .collect();
    result.push_str(&format_section("Flag(s)", &flag_entries));

    result
}

pub fn format_command_list(commands: &HashMap<String, CommandDefinition>) -> String {
    let max_name_length = commands.iter().fold(
        0,
        |acc, (name, _)| {
            if name.len() > acc { name.len() } else { acc }
        },
    );
    let description_start = INDENT + max_name_length + COL_GAP;

    commands.iter().fold(
        "Available commands:\n".to_string(),
        |mut acc, (name, command)| {
            acc.push_str(&format!(
                "{}{:<width$}{}",
                " ".repeat(INDENT),
                name,
                wrap_text(&command.description, description_start),
                width = max_name_length + COL_GAP,
            ));
            acc
        },
    )
}
